lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1629 @@
+	/// Option here.
+	pub last_update: Option<u32>,
+	/// Color assigned to the node
+	pub rgb: [u8; 3],
+	/// Moniker assigned to the node
+	pub alias: [u8; 32],
+	/// Internet-level addresses via which one can connect to the node
+	pub addresses: Vec<NetAddress>,
+	/// An initial announcement of the node
+	//this is cached here so we can send out it later if required by initial routing sync
+	//keep an eye on this to see if the extra memory is a problem
+	pub announcement_message: Option<msgs::NodeAnnouncement>,
+}
+
+impl std::fmt::Display for NodeInfo {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "features: {}, last_update: {:?}, lowest_inbound_channel_fees: {:?}, channels: {:?}", log_bytes!(self.features.encode()), self.last_update, self.lowest_inbound_channel_fees, &self.channels[..])?;
+		Ok(())
+	}
+}
+
+impl Writeable for NodeInfo {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for ref chan in self.channels.iter() {
+			chan.write(writer)?;
+		}
+		self.lowest_inbound_channel_fees.write(writer)?;
+		self.features.write(writer)?;
+		self.last_update.write(writer)?;
+		self.rgb.write(writer)?;
+		self.alias.write(writer)?;
+		(self.addresses.len() as u64).write(writer)?;
+		for ref addr in &self.addresses {
+			addr.write(writer)?;
+		}
+		self.announcement_message.write(writer)?;
+		Ok(())
+	}
+}
+
+const MAX_ALLOC_SIZE: u64 = 64*1024;
+
+impl Readable for NodeInfo {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NodeInfo, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = Vec::with_capacity(cmp::min(channels_count, MAX_ALLOC_SIZE / 8) as usize);
+		for _ in 0..channels_count {
+			channels.push(Readable::read(reader)?);
+		}
+		let lowest_inbound_channel_fees = Readable::read(reader)?;
+		let features = Readable::read(reader)?;
+		let last_update = Readable::read(reader)?;
+		let rgb = Readable::read(reader)?;
+		let alias = Readable::read(reader)?;
+		let addresses_count: u64 = Readable::read(reader)?;
+		let mut addresses = Vec::with_capacity(cmp::min(addresses_count, MAX_ALLOC_SIZE / 40) as usize);
+		for _ in 0..addresses_count {
+			match Readable::read(reader) {
+				Ok(Ok(addr)) => { addresses.push(addr); },
+				Ok(Err(_)) => return Err(DecodeError::InvalidValue),
+				Err(DecodeError::ShortRead) => return Err(DecodeError::BadLengthDescriptor),
+				_ => unreachable!(),
+			}
+		}
+		let announcement_message = Readable::read(reader)?;
+		Ok(NodeInfo {
+			channels,
+			lowest_inbound_channel_fees,
+			features,
+			last_update,
+			rgb,
+			alias,
+			addresses,
+			announcement_message
+		})
+	}
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkMap {
+	/// Known valid channels
+	pub channels: BTreeMap<u64, ChannelInfo>,
```

Comment:
I dont know how I feel about these being pub: RW. We have a few consistency requirements on this and nodes, etc, so maybe it makes more sense to just have accessors for them and options to add new channels directly (without going through the message abstractions)?